sw/device/lib/crypto/drivers/kmac.c

```diff
@@ -550,65 +547,65 @@ static status_t kmac_write_key_block(kmac_blinded_key_t *key) {
- * streaming mode. This decision is in accord with OpenTitan's Crypto Library
- * Specification. Refer to the Hash section of this specification.
- *
- * Current implementation has few limitiations:
+ * streaming mode.
  *
- * 1. `message.data` pointer is assumed to be word-aligned. The case where
- * `data` field is not divisible by 4 is not yet implemented. This is about
- * extra protection against SCA.
+ * This routine does not check input parameters for consistency. For instance,
+ * one can invoke SHA-3_224 with digest_len=32, which will produce 256 bits of
+ * digest. The caller is responsible for ensuring that the digest length and
+ * mode are consistent.
  *
- * 2. Currently, there is no error check on consisteny of the input parameters.
- * For instance, one can invoke SHA-3_224 with digest_len=32, which will produce
- * 256 bits of digest.
- *
- * The caller must ensure that `message` and `digest` have properly
- * allocated `data` fields whose length matches their `len` fields.
+ * The caller must ensure that `message_len` bytes (rounded up to the next 32b
+ * word) are allocated at the location pointed to by `message`, and similarly
+ * that `digest_len` bytes (rounded up to the next 32b word) are allocated at
+ * the location pointed to by `digest`.
  *
  * @param operation The operation type.
  * @param message Input message string.
+ * @param message_len Message length in bytes.
  * @param digest The struct to which the result will be written.
+ * @param digest_len Requested digest length in bytes.
  * @return Error code.
  */
 OT_WARN_UNUSED_RESULT
 static status_t kmac_process_msg_blocks(kmac_operation_t operation,
-                                        crypto_const_uint8_buf_t message,
-                                        crypto_uint8_buf_t *digest) {
-  uint32_t cmd_reg = KMAC_CMD_REG_RESVAL;
-
-  // Assumption(1): `message.len` > 0 and `digest->len` > 0
-  // Assumption(2): message.data % 4 = 0, i.e. the message.data is 32-bit
-  // aligned
-
+                                        const uint8_t *message,
+                                        size_t message_len, uint8_t *digest,
+                                        size_t digest_len) {
+  // Block until KMAC is idle.
   HARDENED_TRY(wait_status_bit(KMAC_STATUS_SHA3_IDLE_BIT, 1));
+
   // Issue the start command, so that messages written to MSG_FIFO are forwarded
   // to Keccak
+  uint32_t cmd_reg = KMAC_CMD_REG_RESVAL;
   cmd_reg = bitfield_field32_write(cmd_reg, KMAC_CMD_CMD_FIELD,
                                    KMAC_CMD_CMD_VALUE_START);
   abs_mmio_write32(kKmacBaseAddr + KMAC_CMD_REG_OFFSET, cmd_reg);
   HARDENED_TRY(wait_status_bit(KMAC_STATUS_SHA3_ABSORB_BIT, 1));
 
-  // Begin by writing a word at a time
-  uint32_t *word_ptr = (uint32_t *)message.data;
+  // Begin by writing a one byte at a time until the data is aligned.
   size_t i = 0;
-  for (; i < message.len / 4; i++) {
+  for (; misalignment32_of((uintptr_t)(&message[i])) > 0; i++) {
```

Comment:
Is there a chance this could result in a fault if the `message_len` is say 2 but the alignment loop results in 3 writes? It is a corner case and not sure what the hardware will do. We could try to test this scenario if we think there is a concern.

Reply:
Good point, I think this is indeed a concern. I'll fix the logic here.